diff --git a/node.js b/node.js
index a3c7ef7e5bfef7e467e166161fc38738234d941e..2859ebb691b3577770215ba528af337e24f6259f 100644
--- a/node.js
+++ b/node.js
@@ -392,12 +392,14 @@ module.exports = {
     var halfYearAgo = Date.now() - TIME_TO_UPDATE_CANIUSE
 
     if (latest !== 0 && latest < halfYearAgo) {
-      console.warn(
-        'Browserslist: caniuse-lite is outdated. Please run:\n' +
-          '  npx update-browserslist-db@latest\n' +
-          '  Why you should do it regularly: ' +
-          'https://github.com/browserslist/update-db#readme'
-      )
+      // this warning is useless, annoying, and should NEVER go
+      // to my stderr. Shut up.
+      // console.warn(
+        // 'Browserslist: caniuse-lite is outdated. Please run:\n' +
+          // '  npx update-browserslist-db@latest\n' +
+          // '  Why you should do it regularly: ' +
+          // 'https://github.com/browserslist/update-db#readme'
+      // )
     }
   },
 