diff --git a/index.d.ts b/index.d.ts
index e5c4acb9f8defb48045a54119a6500ff1a81d4b7..a3c9deedb1d9929085302e69bd0505ec1fd97788 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -101,8 +101,8 @@ declare module "monk" {
     readonly close: () => Promise<void>;
     readonly listCollections: (query?: Object) => Promise<Array<ICollection>>;
 
-    get<T = any>(name: string, options?: CollectionOptions): ICollection<T>;
-    create<T = any>(
+    get<T extends { [key: string]: any } = any>(name: string, options?: CollectionOptions): ICollection<T>;
+    create<T extends { [key: string]: any } = any>(
       name: string,
       creationOption?: CollectionCreateOptions,
       options?: CollectionOptions