diff --git a/index.js b/index.js
index e1c76051045ea9b69b320283a2f27cd2b9ea5cb4..65d183c93ba15242826ccfaf3528e38ebe701f03 100644
--- a/index.js
+++ b/index.js
@@ -65,6 +65,7 @@ class LokiTransport extends Transport {
     }
 
     lokiLabels = Object.assign(lokiLabels, labels)
+    lokiLabels = Object.assign(lokiLabels, rest)
 
     // follow the format provided
     const line = this.useCustomFormat
diff --git a/package-lock.json b/package-lock.json
deleted file mode 100644
index e08faec5f46b8d3802f1a8bf7191fe41d695e047..0000000000000000000000000000000000000000
